File: frontend/src/encoding.js
const INT = 0
const BYTES = 1
const STRING = 2
const BOOL = 3
const ARRAY = 4
const DICT = 5
const SIGNED_INT = 6

class AppendableUint8Array {
  constructor () {
    this.underlying = new Uint8Array(1)
    this.length = 0
  }

  extend () {
    const newUnderlying = new Uint8Array(this.underlying.length * 2)
    newUnderlying.subarray(0, this.underlying.length)
      .set(this.underlying)
    this.underlying = newUnderlying
  }

  appendInt (x) {
    if (this.length + 1 >= this.underlying.length) {
      this.extend()
    }
    this.underlying[this.length] = x
    this.length += 1
  }

  appendArray (array) {
    while (this.length + array.length >= this.underlying.length) {
      this.extend()
    }
    this.underlying.subarray(this.length, this.length + array.length)
      .set(array)

    this.length += array.length
  }

  extract () {
    return this.underlying.slice(0, this.length)
  }
}

function encodeInt (n, appendable) {
  if (n === 0) {
    appendable.appendInt(0)
  } else {
    appendable.appendInt(n % 256)
    encodeInt(n >> 8, appendable)
  }
}

function consumeInt (index, array) {
  let result = 0
  let factor = 0
  while (array[index] !== 0) {
    result = result + (array[index] << 8 * factor)
    factor += 1
    index += 1
  }
  return [index + 1, result]
}

function encodeSignedInt (n, appendable) {
  encodeBool(n < 0, appendable)
  encodeInt(Math.abs(n), appendable)
}

function consumeSignedInt (index, array) {
  let sign, magnitude;
  [index, sign] = consumeBool(index, array);
  [index, magnitude] = consumeInt(index, array)

  return [index, magnitude * (sign ? -1 : 1)]
}

function encodeBytes (array, appendable) {
  encodeInt(array.length, appendable)
  appendable.appendArray(array)
}

function consumeBytes (index, array) {
  let length;
  [index, length] = consumeInt(index, array)
  return [index + length, array.subarray(index, index + length)]
}

const enc = new TextEncoder('utf-8')
const dec = new TextDecoder('utf-8')

function encodeString (string, appendable) {
  return encodeBytes(enc.encode(string), appendable)
}

function consumeString (index, array) {
  let bytes
  [index, bytes] = consumeBytes(index, array)
  return [index, dec.decode(bytes)]
}

function encodeBool (bool, appendable) {
  if (bool) {
    appendable.appendInt(1)
  } else {
    appendable.appendInt(0)
  }
}

function consumeBool (index, array) {
  return [index + 1, array[index] === 1]
}

function encodeArray (array, appendable) {
  encodeInt(array.length, appendable)
  for (let i = 0; i < array.length; i++) {
    encodeAny(array[i], appendable)
  }
}

function consumeArray (index, array) {
  let length, item
  const result = [];

  [index, length] = consumeInt(index, array)
  for (let i = 0; i < length; i++) {
    [index, item] = consumeAny(index, array)
    result.push(item)
  }
  return [index, result]
}

function encodeDict (dict, appendable) {
  // Get number of keys
  let nKeys = 0
  for (const key in dict) {
    if (dict.hasOwnProperty(key)) {
      nKeys += 1
    }
  }

  encodeInt(nKeys, appendable)

  for (const key in dict) {
    if (dict.hasOwnProperty(key)) {
      encodeString(key, appendable)
      encodeAny(dict[key], appendable)
    }
  }
}

function consumeDict (index, array) {
  let nKeys, key, value
  const result = {};

  [index, nKeys] = consumeInt(index, array)
  for (let i = 0; i < nKeys; i++) {
    [index, key] = consumeString(index, array);
    [index, value] = consumeAny(index, array)

    result[key] = value
  }

  return [index, result]
}

function encodeAny (object, appendable) {
  if (typeof object === 'number') {
    appendable.appendInt(SIGNED_INT)
    encodeSignedInt(object, appendable)
  } else if (typeof object === 'string') {
    appendable.appendInt(STRING)
    encodeString(object, appendable)
  } else if (typeof object === 'boolean') {
    appendable.appendInt(BOOL)
    encodeBool(object, appendable)
  } else if (object instanceof ArrayBuffer) {
    object = new Uint8Array(object)
    appendable.appendInt(BYTES)
    encodeBytes(object, appendable)
  } else if (object instanceof Uint8Array) {
    appendable.appendInt(BYTES)
    encodeBytes(object, appendable)
  } else if (object instanceof Array) {
    appendable.appendInt(ARRAY)
    encodeArray(object, appendable)
  } else if (object instanceof Object) {
    appendable.appendInt(DICT)
    encodeDict(object, appendable)
  }
}

function consumeAny (index, array) {
  if (array[index] === INT) {
    return consumeInt(index + 1, array)
  } else if (array[index] === BYTES) {
    return consumeBytes(index + 1, array)
  } else if (array[index] === STRING) {
    return consumeString(index + 1, array)
  } else if (array[index] === BOOL) {
    return consumeBool(index + 1, array)
  } else if (array[index] === ARRAY) {
    return consumeArray(index + 1, array)
  } else if (array[index] === DICT) {
    return consumeDict(index + 1, array)
  } else if (array[index] === SIGNED_INT) {
    return consumeSignedInt(index + 1, array)
  } else {
    throw new Error('invalid type ' + array[index])
  }
}

function encode (dict) {
  const appendable = new AppendableUint8Array()
  encodeAny(dict, appendable)
  return appendable.extract()
}

function decode (array) {
  let index = 0
  let result

  [index, result] = consumeAny(index, array)

  return result
}

export default {encode, decode}

File: backend/encoding.py
INT = 0
BYTES = 1
STRING = 2
BOOL = 3
ARRAY = 4
DICT = 5
SIGNED_INT = 6

B_INT = b'\x00'
B_BYTES = b'\x01'
B_STRING = b'\x02'
B_BOOL = b'\x03'
B_ARRAY = b'\x04'
B_DICT = b'\x05'
B_SIGNED_INT = b'\x06'

def encode_int(n):
    if n == 0:
        return b'\x00'
    else:
        return bytes([n % 256]) + encode_int(n // 256)

def consume_int(index, bstring):
    result = 0
    factor = 1
    while bstring[index] != 0:
        result += bstring[index] * factor
        factor *= 256
        index += 1
    return index + 1, result

def encode_signed_int(n):
    return encode_bool(n < 0) + encode_int(abs(n))

def consume_signed_int(index, bstring):
    index, sign = consume_bool(index, bstring)
    index, magnitude = consume_int(index, bstring)

    return index, magnitude * (-1 if sign else 1)

def encode_bytes(string):
    return encode_int(len(string)) + string

def consume_bytes(index, bstring):
    index, length = consume_int(index, bstring)
    result = bstring[index : index + length]

    return index + length, result

def encode_string(string):
    return encode_int(len(string)) + string.encode()

def consume_string(index, bstring):
    index, bstr = consume_bytes(index, bstring)
    return index, bstr.decode()

def encode_bool(b):
    return b'\x01' if b else b'\x00'

def consume_bool(index, bstring):
    return index + 1, (bstring[index] == 1)

def encode_array(array):
    return encode_int(len(array)) + b''.join(encode_any(obj) for obj in array)

def consume_array(index, bstring):
    index, length = consume_int(index, bstring)
    arr = []

    for i in range(length):
        index, item = consume_any(index, bstring)
        arr.append(item)

    return index, arr

def encode_dict(d):
    result = b''
    result += encode_int(len(d))
    for key in d:
        result += encode_string(key)
        result += encode_any(d[key])
    return result

def consume_dict(index, bstring):
    index, length = consume_int(index, bstring)

    result = {}
    for i in range(length):
        index, key = consume_string(index, bstring)
        index, value = consume_any(index, bstring)
        result[key] = value

    return index, result

def encode_any(data):
    if type(data) == int:
        return B_SIGNED_INT + encode_signed_int(data)
    elif type(data) == bytes:
        return B_BYTES + encode_bytes(data)
    elif type(data) == bool:
        return B_BOOL + encode_bool(data)
    elif type(data) == str:
        return B_STRING + encode_string(data)
    elif type(data) == list:
        return B_ARRAY + encode_array(data)
    elif type(data) == dict:
        return B_DICT + encode_dict(data)

def consume_any(index, bstring):
    if bstring[index] == INT:
        return consume_int(index + 1, bstring)
    elif bstring[index] == BYTES:
        return consume_bytes(index + 1, bstring)
    elif bstring[index] == STRING:
        return consume_string(index + 1, bstring)
    elif bstring[index] == BOOL:
        return consume_bool(index + 1, bstring)
    elif bstring[index] == ARRAY:
        return consume_array(index + 1, bstring)
    elif bstring[index] == DICT:
        return consume_dict(index + 1, bstring)
    elif bstring[index] == SIGNED_INT:
        return consume_signed_int(index + 1, bstring)

def encode(obj):
    return encode_any(obj)

def decode(bstring):
    _, result = consume_any(0, bstring)
    return result
